zebra-consensus/src/checkpoint/tests.rs

```diff
@@ -238,91 +243,120 @@ async fn continuous_blockchain() -> Result<(), Report> {
         .collect();
 
-    let mut checkpoint_verifier =
-        CheckpointVerifier::from_list(checkpoint_list).map_err(|e| eyre!(e))?;
-
-    // Setup checks
-    assert_eq!(
-        checkpoint_verifier.previous_checkpoint_height(),
-        BeforeGenesis
-    );
-    assert_eq!(
-        checkpoint_verifier.target_checkpoint_height(),
-        WaitingForBlocks
-    );
-    assert_eq!(
-        checkpoint_verifier.checkpoint_list.max_height(),
-        BlockHeight(10)
-    );
-
-    let mut handles = FuturesUnordered::new();
-
-    // Now verify each block
-    for (block, height, _hash) in blockchain {
-        /// SPANDOC: Make sure the verifier service is ready
-        let ready_verifier_service = checkpoint_verifier
-            .ready_and()
-            .map_err(|e| eyre!(e))
-            .await?;
-
-        /// SPANDOC: Set up the future for block {?height}
-        let verify_future = timeout(
-            Duration::from_secs(VERIFY_TIMEOUT_SECONDS),
-            ready_verifier_service.call(block.clone()),
-        );
-
-        /// SPANDOC: spawn verification future in the background
-        let handle = tokio::spawn(verify_future.in_current_span());
-        handles.push(handle);
-
-        // Execution checks
-        if height < checkpoint_verifier.checkpoint_list.max_height() {
+    /// SPANDOC: Verify blocks, restarting at {?restart_height}
+    {
+        let initial_tip = restart_height
+            .map(|BlockHeight(height)| &blockchain[height as usize].0)
+            .cloned();
+        let mut checkpoint_verifier =
+            CheckpointVerifier::from_list(checkpoint_list, initial_tip).map_err(|e| eyre!(e))?;
+
+        // Setup checks
+        if restart_height
+            .map(|h| h >= checkpoint_verifier.checkpoint_list.max_height())
+            .unwrap_or(false)
+        {
+            assert_eq!(
+                checkpoint_verifier.previous_checkpoint_height(),
+                FinalCheckpoint
+            );
             assert_eq!(
                 checkpoint_verifier.target_checkpoint_height(),
-                WaitingForBlocks
+                FinishedVerifying
             );
         } else {
             assert_eq!(
                 checkpoint_verifier.previous_checkpoint_height(),
-                FinalCheckpoint
+                restart_height.map(InitialTip).unwrap_or(BeforeGenesis)
             );
             assert_eq!(
                 checkpoint_verifier.target_checkpoint_height(),
-                FinishedVerifying
+                WaitingForBlocks
             );
         }
-    }
+        assert_eq!(
+            checkpoint_verifier.checkpoint_list.max_height(),
+            BlockHeight(9)
+        );
 
-    while let Some(result) = handles.next().await {
-        result??.map_err(|e| eyre!(e))?;
-    }
+        let mut handles = FuturesUnordered::new();
+
+        // Now verify each block
+        for (block, height, _hash) in blockchain {
```

Comment:
Wondering if this could be expressed as a `zebra_test::Transcript`

It seems like the only thing that's missing is some sort of customizable callback to do additional assertions after each request. I think we might be able to bring in the assertions as an additional service wrapping the transcript.

Reply:
Would you like to open another ticket?

I'm not sure where to start here.

Reply:
Yea, I'll open another issue, no point holding this up.